### ISSUE TITLE
Let's call them initializers

### ISSUE THREAD

**Issue by mikdusan** [ziglang/zig#7764]
This is a proposal to change docs for the following terms:

- anonymous struct literal
- anonymous list literal
- anonymous union literal
- tuple

TL;DR Simplify. For. Idiomatic. Usage.

Why? Because 99% of the time when new users, existing users, use these terms they don't care about "anonymous" or "literal" part of it. Let's relax things and **simplify** for idiomatic usage, ~~not precision for mad-scientist locked away in a rubber-room lab coding 50 or 60 compiler passes~~. Ok that last part was a bit much.

Related is #7758 which proposes language-level syntax changes to alleviate ambiguity. I believe we do not need those changes if this PR is accepted as new users will graps simplified terms faster.

edit: there is **accepted** proposal #4335 which impacts the concept of tuple in the language, and this proposal

### Basic terminology used here:

- "**literal**" is a thing with lexical-specified value
- "**anonymous**" is a thing with no name

### What is an "**anonymous struct literal**"?
```zig
_ = .{
    .out = "_out/",
    .bin = "_out/bin/",
    .lib = "_out/lib/",
    .cache = "_out/zig-cache/",
};
```

- Q. Is it anonymous? yes.
- Q. Is it a struct? yes.
- Q. Is it a literal? yes.

### What happens if we use runtime data?

```zig
const complete_literal = .{
    .out = "_out/",
    .bin = "_out/bin/",
    .lib = "_out/lib/",
    .cache = "_out/zig-cache/",
};

var custom_out: []const u8 = getCustomOut();
const partial_literal = .{
    .out = custom_out,
    .bin = "_out/bin/",
    .lib = "_out/lib/",
    .cache = "_out/zig-cache/",
};
```

Literals are no longer complete if mixed with runtime values. This is an important distinction for later.

### What do we call this?

```zig
_ = .{};
```

It's an empty thing. But an empty what? struct, list or union?

### Idiomatic Usage

The common use of anonymous struct literals are as initializers. And for that reason I think we should coin the term "struct initializer". Drop "anonymous" because it's idiomatically unimportant. And we no longer care if it is a literal and if it is complete or not.

It also plays nice when describing a named-type initializer vs "struct initializer" because in that equation only the named-type has a name and saying "anonymous struct initializer" would be superfluous.

```zig
const Foo = struct {
    bar: bool = false,
};

// ok: use a Foo initializer
var a = Foo{};

// use a struct initializer
// superfluous: use an anonymous struct initializer
var b: Foo = .{};
```

### How does the term "initializer" pan out?

Let's see how things play out by calling these things initializers.

#### "empty initializer"

```zig
_ = .{};

var thing1: struct {
    first: []const u8 = "",
    last: []const u8 = "",
} = undefined;
thing1 = .{};

var thing2: [0]u32 = undefined;
thing2 = .{};

// this might be a comptiler-bug allowing me to do this?
var thing3: union(enum) {
    big: u64 = undefined,
    tiny: u8,
} = undefined;
thing3 = .{};
```

#### "struct initializer"

```zig
var thing1: struct {
    first: []const u8 = "",
    last: []const u8 = "",
} = undefined;
thing1 = .{
    .first = "Gordonhello",
    .last = "Gecko",
};
```

#### "array initializer"

```zig
var thing2: [2]u32 = undefined;
thing2 = .{ 42, 88 };
```

#### "union initializer"

```zig
var thing3: union(enum) {
    big: u64,
    tiny: u8,
} = undefined;
thing3 = .{ .tiny = 0 };
```

### "tuple"

I found 1 use of this term in zig langref. It is used to to describe all 3 aonymous literals { struct, array, union } and we do not have a language-level type `tuple` that can be used as such. It should be dropped.

### Conclusion

1. re-term "anonymous struct literal" → "struct initializer"
2. re-term "anonymous list literal" → "list initializer"
3. re-term "anonymous union literal" → "union initializer"
4. introduce the term "empty initializer"
5. drop the term "tuple"

**Comment by EleanorNB** [comment]
I like this. It also makes them seem less like actual types, which they're not really -- I've discussed with @SpexGuy in the past about relaxing all layout requirements, and having them essentially just be a way to refer to multiple values pending a result location. Not sure if that will go through, but in any case I would certainly like to be rid of them in the type system.

**Comment by mikdusan** [comment]
> edit: adding accepted proposal #4335 which impacts the concept of tuple in the language, and this proposal

just putting this edit here too.

**Comment by EleanorNB** [comment]
One point on naming: people are going to use short names anyway, CIP the current discussion on Discord where everyone is still referring to anonymous list literals as "tuples", as well as anonymous struct literals which really don't fit that definition. So, I propose the following short names:
- struct initialiser -> "record"
- array initialiser -> "tuple"
"Empty initialiser" I think is fine, since it's only ever really used for that purpose, and "empty record/tuple" isn't all that misleading unlike the above ambiguity.

**Comment by mikdusan** [comment]
I do like the usage of `record` ... so old school. I don't even know why we rarely see it these days. It's so appropriate here because it really is just data.

**Comment by fengb** [comment]
I don't like `list` because it's often relates to arrays and implies a homogeneous collection. A `tuple` makes much more sense in my brain.

@EleanorNB https://github.com/ziglang/zig/issues/4335 will make tuple definitions making them not just initialization

**Comment by mikdusan** [comment]
This is a little bike shedding but gonna sound it out.
<img width="115" alt="image" src="https://user-images.githubusercontent.com/48141240/104466905-b757e680-5583-11eb-8d4b-363718f8d4bf.png">


→ record: `.{ .one=1, .two=2 }`
→ tuple: `.{ .one, .two, .there }`
→ empty initializer: `.{}`
→ empty: `.{}` // just gonna try this on

"you need to initialize the struct with a record"
"you need to initialize the union with a record"
"you need to initialize Foo with a record"
"you need to initialize Foo with an empty"
"you need to initialize the array with a tuple"
"you need to initialize Bar with a tuple"
"you need to initialize Bar with an empty"

"Foo can be initialized with a record"
"Foo accepts an empty initializer"
"Foo accepts an empty"
"Foo can be initialized with an empty"

**Comment by mikdusan** [comment]
@fengb `empty definition` is nice. This might be the trifecta: `empty definition`, `record`, `tuple`

**Comment by EleanorNB** [comment]
Then it's a two-word identifier, which would either need a conspicuous join or (*shudder*) an at-quote. Regardless of doc name, the `typeInfo` name should be `empty`.

Regarding doc name, "definition" to me means type declaration -- that's what it *is*, the dot-brace is what it *has*. This thing *initialises*, so in my mind it should be an initialiser. I'm not attached to this opinion though.

@fengb Sure, but that doesn't mean it needs a defined *layout*, or even necessarily be contiguous. Actually, come to think of it, isn't the main use of type annotations on bare initialisers for varargs, which can't use them anyway?

**Comment by mikdusan** [comment]
musing a single syntax change:

```zig
_ = .{}; // empty record
_ = .[]; // empty tuple

var a: [2]u32 = .[ 42, 88 ];
var b: [2]u32 = .{ 42, 88 ]; // compile error: cannot initialize array with record
```

**Comment by EleanorNB** [comment]
My instinct is no. The dot in dotbrace stands for an inferred type -- replacing it with a concrete type should yield the same result. So, an explicitly typed array literal would look like `[3]u8[1, 2, 3]`. I don't know if this is ambiguous, but it's certainly ugly.

**Comment by fengb** [comment]
> @fengb Sure, but that doesn't mean it needs a defined layout, or even necessarily be contiguous. Actually, come to think of it, isn't the main use of type annotations on dotbraces for varargs, which can't use them anyway?

Existing usecases maybe but I have these other changes I'd like to do:

```zig
fn foo(args: *extern tuple{ i32, i64 }) void {}
fn bar(args: *extern tuple{ f32, f64 }) void {}

if (cond) {
    foo(@ptrCast(*extern tuple{ i32, i64 }, mem.ptr));
} else {
    bar(@ptrCast(*extern tuple{ f32, f64 }, mem.ptr));
}
```

I'm worried we're hijacking tuple to mean anonymous "list" literal, whereas the internal compiler definition has nothing to do with being anonymous.

**Comment by EleanorNB** [comment]
Platform ABIs don't even *define* tuples, and if they did they'd follow the same rules as structs. `extern tuple` makes no sense, and we can't make it make sense.

"Tuple" (under this proposal) means "dotbrace literal without labels". There can be named dotbraces as well as anonymous typed literals. I don't see how we're assuming "anonymous", unless we're also disabling RLS for this case, which is not being proposed and nor should it.

**Comment by Mouvedia** [comment]
I agree with 1, 3 and 4. I am not so sure about 3 and 5.

@mikdusan What's the difference between "list initializer" and "array initializer"?
BTW [it can also be spelled initialiser](https://en.wiktionary.org/wiki/initialiser#Noun).

**Comment by mikdusan** [comment]
>What's the difference between "list initializer" and "array initializer"?

@Mouvedia, in consideration of tuple #4335, I would probably go with Eleanor's suggestion and use the terms "record" and "tuple" and "empty initializer", and I believe that avoids the "list" vs "array" per Feng's point (paraphrasing) a tuple is not required to be of homogeneous type.

##### (note: I should open a competing proposal with this in mind)

**Comment by mikdusan** [comment]
I am closing this with no intended further action needed.

With all the valuable input in this thread it has convinced me that status-quo terminology is not as much of an issue as I had  originally thought and changing it is not compelling enough.

Thank you all for your input!
